internal/controller/apiextensions/composition/reconciler.go

```diff
@@ -0,0 +1,175 @@
+)
+
+const (
+	timeout = 2 * time.Minute
+)
+
+// Error strings
+const (
+	errGet       = "cannot get Composition"
+	errListRevs  = "cannot list CompositionRevisions"
+	errCreateRev = "cannot create CompositionRevision"
+)
+
+// Event reasons.
+const (
+	reasonCreateRev event.Reason = "CreateRevision"
+)
+
+// Setup adds a controller that reconciles Compositions by creating new
+// CompositionRevisions for each revision of the Composition's spec.
+func Setup(mgr ctrl.Manager, log logging.Logger) error {
+	name := "revisions/" + strings.ToLower(v1.CompositionGroupKind)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1.Composition{}).
+		Owns(&v1alpha1.CompositionRevision{}).
+		Complete(NewReconciler(mgr,
+			WithLogger(log.WithValues("controller", name)),
+			WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+// ReconcilerOption is used to configure the Reconciler.
+type ReconcilerOption func(*Reconciler)
+
+// WithLogger specifies how the Reconciler should log messages.
+func WithLogger(log logging.Logger) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.log = log
+	}
+}
+
+// WithRecorder specifies how the Reconciler should record Kubernetes events.
+func WithRecorder(er event.Recorder) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.record = er
+	}
+}
+
+// NewReconciler returns a Reconciler of Compositions.
+func NewReconciler(mgr manager.Manager, opts ...ReconcilerOption) *Reconciler {
+	kube := unstructured.NewClient(mgr.GetClient())
+
+	r := &Reconciler{
+		client: kube,
+		log:    logging.NewNopLogger(),
+		record: event.NewNopRecorder(),
+	}
+
+	for _, f := range opts {
+		f(r)
+	}
+	return r
+}
+
+// A Reconciler reconciles Compositions by creating new CompositionRevisions for
+// each revision of the Composition's spec.
+type Reconciler struct {
+	client client.Client
+
+	log    logging.Logger
+	record event.Recorder
+}
+
+// Reconcile a Composition.
+func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("request", req)
+	log.Debug("Reconciling")
+
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	comp := &v1.Composition{}
+	if err := r.client.Get(ctx, req.NamespacedName, comp); err != nil {
+		log.Debug(errGet, "error", err)
+		r.record.Event(comp, event.Warning(reasonCreateRev, err))
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), errGet)
+	}
+
+	if meta.WasDeleted(comp) {
+		return reconcile.Result{}, nil
+	}
+
+	currentHash := comp.Spec.Hash()
+
+	log = log.WithValues(
+		"uid", comp.GetUID(),
+		"version", comp.GetResourceVersion(),
+		"name", comp.GetName(),
+		"spec-hash", currentHash,
+	)
+
+	rl := &v1alpha1.CompositionRevisionList{}
+	if err := r.client.List(ctx, rl, client.MatchingLabels{v1alpha1.LabelCompositionName: comp.GetName()}); err != nil {
+		log.Debug(errListRevs, "error", err)
+		r.record.Event(comp, event.Warning(reasonCreateRev, err))
```

Comment:
nit: shouldn't the event also include information that this error happened during listing revisions (same for get above)? 

For example:
```suggestion
		r.record.Event(comp, event.Warning(reasonCreateRev, errors.Wrap(err, errListRevs)))
```